### ISSUE TITLE
Compiler not optimizing away multiple by zero

### ISSUE THREAD

**Issue by deanm** [HaxeFoundation/haxe#5763]
Thanks to inlining I get some pretty straight matrix multiplication code, but for something standard like gluPerspective where you have a lot of zeros in the matrix, Haxe generates JavaScript like:

	self.a11 = a11 * b11 + a12 * 0 + a13 * 0 + a14 * 0;
	self.a12 = a11 * 0 + a12 * f + a13 * 0 + a14 * 0;
	self.a13 = a11 * 0 + a12 * 0 + a13 * b33 + a14 * -1;
	self.a14 = a11 * 0 + a12 * 0 + a13 * b34 + a14 * 0;
	self.a21 = a21 * b11 + a22 * 0 + a23 * 0 + a24 * 0;
	self.a22 = a21 * 0 + a22 * f + a23 * 0 + a24 * 0;
	self.a23 = a21 * 0 + a22 * 0 + a23 * b33 + a24 * -1;
	self.a24 = a21 * 0 + a22 * 0 + a23 * b34 + a24 * 0;
	self.a31 = a31 * b11 + a32 * 0 + a33 * 0 + a34 * 0;
	self.a32 = a31 * 0 + a32 * f + a33 * 0 + a34 * 0;
	self.a33 = a31 * 0 + a32 * 0 + a33 * b33 + a34 * -1;
	self.a34 = a31 * 0 + a32 * 0 + a33 * b34 + a34 * 0;
	self.a41 = a41 * b11 + a42 * 0 + a43 * 0 + a44 * 0;
	self.a42 = a41 * 0 + a42 * f + a43 * 0 + a44 * 0;
	self.a43 = a41 * 0 + a42 * 0 + a43 * b33 + a44 * -1;
	self.a44 = a41 * 0 + a42 * 0 + a43 * b34 + a44 * 0;

Any reason the multiply by zeros can't go drop out, and the var * -1 can't become -var?

**Comment by kaikoga** [comment]
I suppose it's because NaN * 0 = NaN

**Comment by deanm** [comment]
Yeah, I had thought about the case of NaN.  Many compilers offer something like -ffast-math which will make optimizations assuming things are never NaN, etc.  I think it would be really nice to have something like that, looking at the expression above it's really clear that you want the compiler to do something smarter with it, or I have to do it by hand...

**Comment by ncannasse** [comment]
We're trying to avoid such global flags that affects the language semantics since it might trigger some errors in libraries accidentally (or not) relying on this particular behavior.

**Comment by deanm** [comment]
Anything we can do on a local class / function level?  It's not really something that easily addressed by a macro is it, because it my case it's a byproduct of inlining, so I would need a macro that basically runs on the code post optimization...

**Comment by DevRubicate** [comment]
+1 for some sort of solution to this.

It's quite frustrating, not just because of the unnecessary arithmetic, but also because the unnecessary operations and variable assignments that stays around for the multiplication of zero, ends up blocking other optimizations further down along the code, creating a sort of avalanche effect. I have a bunch of mathematical matrices, and hundreds of lines of mathematical operations get unraveled and optimized into a neat one-liner when I help Haxe along by adding a conditional (value != 0), but that introduces a branch which is bad in other ways.

I understand that this isn't really a bug, NaN is actually supposed to behave like this even when multiplying by zero. Might it be possible to have something on a local scope, or maybe a new abstract primitive number type for fastmath which dares make more assumptions?

**Comment by RealyUniqueName** [comment]
We can probably introduce a meta on a field or expression level.
E.g.
```haxe
@:fastMath function method() { ... }
```
or
```haxe
function method() {
    normalMath();
    @:fastMath {
        ...
    }
}
```

**Comment by DevRubicate** [comment]
I'm not sure if this will be useful, however, in vanilla JS, one way to get faster math without this particular pitfall is to add "|0" to all values as part of the expression. I believe asm.js does this, and it ensures that stuff like null, undefined and NaN gets treated as a zero (as a bonus it can also make JS execution go faster as the runtime engine dares make more assumptions). In essence it allows JS to "dumb down" it's overloaded Number type into a more primitive integer, ensuring it has no floats, no infinity, no null values, no nothing except 32 bits holding an integer. Even (null|0) becomes 0.

So this:

```
a * b
```

Would be written as this:

```
(a|0) * (b|0)
```

If variable "b" here is 0, we can be 100% confident that the result of the expression would also be 0 during runtime, even if "a" contains NaN. That way, you can ensure that the runtime evaluation of something like that is identical to the compiler evaluation, the compiler doesn't have to worry about it's optimizations producing the wrong answer.

Obviously the real question is what would be the correct way to turn this feature on. I'm in agreement with @ncannasse that having a global flag for fastmath is a bad idea, NaN isn't some glitch or bug, it's a genuine part of the JS runtime, and in some contexts it makes sense to have it around. The dev might need it somewhere else, and libraries might utilize it.

The two most realistic ideas I can see are either to have some sort of "fastMath block" like @RealyUniqueName is suggesting, or to have a secondary abstract integer core type called something like FastInt where the compiler knows it's safe to optimize, and adds |0 during compilation to ensure that runtime behaves the same way (hopefully other targets than JS have similar tricks?).

**Comment by nadako** [comment]
I was actually (silently) wondering for quite some time now about whether we should indeed add `|0` for integer operations as this guarantees integer behaviour and I also _heard_ that it helps JS engines to figure out what are they dealing with at run time.

**Comment by nadako** [comment]
It would be cool to have some benchmarks comparison on all the major JS engines (not sure how easy it is to trick the JIT to _not_ always figure it out)
